Step2 구현, step1 피드백 수정

dahun-lee-daji opened the pull request:
- step2 를 마치고 PR드립니다.

이번 미션에서는 구현은 어렵지 않았지만, github가 발목을 너무 오랫동안 잡고있었습니다. 몇 시간 동안 삽질하다가 local repository가 만신창이가 되어버려서, 되돌릴 수가 없어서 밀어버리고 새로 만들었습니다....

feedback 사항 : 
1. AppDelegate, SceneDelegate 내부에 불필요해보이는 함수를 삭제하는데, 함수 내에 내용이 없으므로 삭제해도 괜찮을거라 생각했습니다. 물론, 자료조사도 했지만 유의미하지는 않았습니다. 불안하여 삭제하고 빌드하고 몇 번 반복하며 해당 내용을 수정했습니다.
2. 혼란을 줄 수 있는 fatalError() -> personalError()로 변경했습니다.
3. 변수, 상수 선언 수정했습니다.
4. 함수 네이밍 규칙을 다시 읽고 반영해봤습니다.

요구사항 구현 : 
class PlayingCard를 구현했습니다.
규격에 맞는 랜덤 카드 생성, 원하는 모양과 숫자의 카드 생성을 구현했고, console에 표시 할 수 있습니다.

dahun-lee-daji committed:
step1 피드백 반영

appdelegate, scenedelegate 불필요한 내용 삭제
혼동을 피하기 위해 class TroubleShooter 내 fatalError -> personalError
변수, 상수 선언 수정
함수 네이밍 규칙 반영

dahun-lee-daji committed:
step2 요구사항 구현

class PlayingCard 구현
console에 카드 내용 표시 -> random 1, heartQ, spade7

dahun-lee-daji committed:
Update README.md

godrm commented on PokerGameApp/PokerGameApp/PlayingCard.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+    
+    func showCard() {
+        print(suit.rawValue, rank[number] ?? "nil", separator : "" )
```
값을 보관하거나 처리하는 타입 내부에서 직접 출력하는 경우는 없습니다. 
그 타입을 호출하는 상위 모듈에서 출력 방식을 결정하도록 값을 전달하는 게 좋습니다. 
CustomStringConvertible 프로토콜에 대해 학습해보고 Card에 채택해서 적용해보세요.

godrm commented on PokerGameApp/PokerGameApp/PlayingCard.swift:
```diff
@@ -0,0 +1,39 @@
+    
+    init() {
+        self.suit = Suits.allCases.randomElement()!
+        self.number = Int.random(in: 1...13)
```
PlayingCard는 하나의 카드를 표현하는 혹은 모델링하는 타입이어야 합니다.
이렇게 랜덤값을 갖고 있으면 원하는 카드가 생성될 때까지 반복해야 하는걸까요? 
특정한 조건의 데이터를 생성하고 보관하는 역할과 그것을 랜덤하게 다루는 로직을 분리해보세요. 서로 다른 역할입니다.

dahun-lee-daji replied:
JK, 제가 이해한 내용이 맞는지 확인 부탁드려요!
저는 특정한 조건의 카드와 랜덤한 카드를 만드는 두개의 init을 만들었습니다.
하지만, 랜덤한 카드를 만드는 init은 적합하지 않으며, 차라리 특정한 조건의 카드를 만들고, 이 카드를 랜덤하게바꾸는게 더 낫다고 알려주시는 건가요??😃

godrm replied:
아.. init()이 두 개였군요. 
위에 init은 init 이라기 보다는 팩토리 메소드에 가깝습니다. 
init 이 아니라 `PlayingCard.init(랜덤값)`을 호출해서 객체를 생성하는 상위 모듈의 동작에 가깝다는 겁니다.
init()은 내부 속성을 초기화하는 것에 그 목표가 있습니다. 그래서 서로 다른거라는 의미였습니다. 
만약 두 개 init을 모두 쓰고 싶다면 위에 init을 convenience init으로 만들어서 self.init()을 호출하는 구조여야 합니다.

dahun-lee-daji replied:
이해했습니다. 감사합니다!

godrm reviewed (changes requested):
PR 열어놓겠습니다. 다음 두 가지 피드백을 개선해서 다시 PUSH해보세요.
모두 수정하고 댓글을 남겨주세요.

dahun-lee-daji committed:
피드백 사항 반영

값을 보관,처리하는 타입 내부에서 직접 출력하면 안되므로, Suits : CustomStringConvertible을 채택, showCard()를 수정.
Factory method style의 랜덤 카드 init()을 convenience init으로 수정

dahun-lee-daji commented:
@godrm 피드백 사항 반영 완료했습니다. 😊

godrm commented on PokerGameApp/PokerGameApp/PlayingCard.swift:
```diff
@@ -7,33 +7,48 @@
+    
     func showCard() {
-        print(suit.rawValue, rank[number] ?? "nil", separator : "" )
+        print(suit, rank[number] ?? "nil", separator : "" )
```
이 메소드는 의미가 없지만 일단 넘어갑시다.

godrm commented on PokerGameApp/PokerGameApp/PlayingCard.swift:
```diff
@@ -7,33 +7,48 @@
     
+}
+
+class PlayingCard {
```
그리고보니 PlayingCard에 CustomStringConvertible가 없네요? 
showCard() 대신에 이게 필요한데요?

dahun-lee-daji replied:
Suits에 추가했는데, 말씀하신 내용대로 PlayingCard에 들어가는게 더 적합해 보이네요, 다시 수정하겠습니다!

dahun-lee-daji committed:
customStringConvertible 적용 대상 수정

적용 대상 : Suits -> PlayingCard
이에 따라서, 불필요한 showCard() 메서드 삭제

dahun-lee-daji commented:
앞으로는 피드백 사항을 심사숙고해서 적용하겠습니다.
지적해주셔서 감사합니다!

dahun-lee-daji committed:
Suits에 CustomStringConvertible 재 추가

PlayingCard에 추가되었으므로 필요 없을 것이라 생각했지만, 코드 단순화를 위해 다시 추가

godrm commented on PokerGameApp/PokerGameApp/PlayingCard.swift:
```diff
@@ -7,18 +7,26 @@
-        case heart = "\u{2665}"
-        case diamond = "\u{2666}"
-        case club = "\u{2663}"
+        return "\(suit)\(rank[number] ?? "nil")"
```
이상하지 않나요? Suits는 단순해 졌는데 rank 처리만 길어졌으니까
Rank에도 CustomStringConvertible를 채택하면 되지 않을까요? 
이 부분도 고민해보세요.

dahun-lee-daji replied:
<img width="400" alt="스크린샷 2021-02-17 오후 7 46 14" src="https://user-images.githubusercontent.com/68788135/108193491-d801f600-7158-11eb-8a4c-9e93a1f11064.png">

Dictionary ( rank )는 이미 CustomStringConvertible을 준수하고 있습니다. rank의 처리가 긴 것은 옳지 않은 key에 대한 오류 처리 내용에 가깝다고 생각합니다.

`
extension Dictionary : CustomStringConvertible where Key == Int, Value == String {
    var description : String {}
}
`

위와 같은 내용으로 시도했었는데, 더 이상의 진전이 없는 것 같습니다,,,

godrm replied:
아.. 타입을 선언하지 않았군요. 다음 단계에서는 Dictionary<> 대신에 타입을 직접 선언해보세요.
